### PR TITLE
save/load improvements

### PR DESCRIPTION
**Pull request by ianspektor**
save
- [x] restrict `save` to only receiving node arguments and returning dict of nodes
- [x] receive parameters directly instead of as a dict in `save`

load
- [x] rename load to load_graph
- [x] create `load` function to restore a saved function
- [x] dynamically generate signature of loaded function
- [x] allow loaded function to receive positional params

follow up work / to discuss
- fixing static typing of loaded function (might not be possible, same issue as the decorator, per definition can't fix static types dynamically)

### CODE REVIEW

File: temporian/core/serialization.py [hunk 1]
```diff
@@ -55,95 +56,114 @@
...
+        kwargs = _construct_kwargs_from_args_and_kwargs(
+            list(g.named_inputs.keys()), args, kwargs
         )
-    if isinstance(input, Schema):
-        return create_node_with_new_reference(schema=input)
-    if isinstance(input, EventSet):
-        return input.node()
-    if isinstance(input, EventSetNode):
-        return input
+        return g.apply_on_inputs(named_inputs=kwargs)
```
Comment (achoum) [inline temporian/core/serialization.py:144]: This is what append when we mix functional and graph code :)

Nit: Can you move the "list(g.named_inputs.keys())" outside of the "fn" ?

File: temporian/core/graph.py [hunk 2]
```diff
@@ -161,6 +163,44 @@ def p3(title, l):
...
+
+        for name, new_node in named_inputs.items():
+            if name not in g.named_inputs:
+                raise ValueError(
+                    f"Input node {name} is not in the graph's inputs. Inputs:"
+                    f" {g.named_inputs}"
+                )
+            old_node = g.named_inputs[name]
+            # Replace node as input in all operators that depend on it
+            for operator in g.operators:
```
Comment (achoum) [inline temporian/core/graph.py:198]: Can you add a TODO to create and manage a mapping from named_input to operator in the graph to maje this method more efficient?

**Review by achoum — COMMENTED** [review]
It looks good :)

File: temporian/core/graph.py [hunk 3]
```diff
@@ -161,6 +163,44 @@ def p3(title, l):
...
+            named_inputs: The new inputs to the graph.
+
+        Returns:
+            The graph's named outputs.
+        """
+
+        # Avoid messing with the objects in the received graph
+        # Else each time we apply the graph on inputs it changes
+        # Might be possible to optimize by copying only what's necessary
+        g = deepcopy(self)
```
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/core/graph.py:185]: great catch! will write a test for it too

File: temporian/core/graph.py [hunk 4]
```diff
@@ -161,6 +163,44 @@ def p3(title, l):
             p3("Output", self.outputs)
         return s
 
+    def apply_on_inputs(
```
[the remark this answers is not in the document]
Comment (ianspektor) [reply temporian/core/graph.py:166]: I had changed it to that but changed my mind: "replace_named_inputs" makes me feel we're returning a graph with the replaced inputs, but that's not the case, we're actually applying the graph to new nodes and returning the outputs. wdyt?

[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/graph.py:166]: Sg

File: temporian/core/serialization.py [hunk 5]
```diff
@@ -248,6 +268,52 @@ def load(
     return inputs, outputs
 
 
+def _process_fn_input(input: Any) -> EventSetNode:
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/serialization.py:275]: You are right, this is more compact.

Could there be a risk for the user to miss-understand what the function is doing when feeding an EventSet? If the user sees that a function can take three types of inputs, maybe the user will think there should be a difference. If there is no such risk, it is okay to have multiple version.

I'll let you decide.

Nit: If you add support for Schema, can you create the node with tp.input_node? The "create_node_with_new_reference" is really only designed for operator definition.

File: temporian/core/test/serialization_test.py [hunk 6]
```diff
@@ -163,109 +175,269 @@ def test_serialize_attributes(self):
...
+            "y: temporian.core.data.node.EventSetNode, "
+            "z: temporian.core.data.node.EventSetNode"
+            in doc
+        )
+        self.assertTrue(
+            "-> Dict[str, temporian.core.data.node.EventSetNode]" in doc,
+        )
 
 
 if __name__ == "__main__":
```
Comment (achoum) [inline temporian/core/test/serialization_test.py:443]: Can you also the the following scenarios:

Function compositions
===================

@tp.compile
def f(x: tp.EventSetOrNode):
    return tp.prefix("f_",x)
    
@tp.compile
def g(x: tp.EventSetOrNode):
    return tp.prefix("g_",x)
    
@tp.compile
def h(x: tp.EventSetOrNode):
    return f(g(x))
    
tp.save(f, ...)

Partial function
=============

@tp.compile
def f(x: tp.EventSetOrNode, b: bool):
    if b:
        return tp.prefix("true_",x)
    else:
        return tp.prefix("false_",x)
    
tp.save(partial(f,b=True), ...)

and

Recursive function
==============

@tp.compile
def f(x: tp.EventSetOrNode, n: int=5):
    if n>0:
        return f(tp.prefix(f"{n}_",x), n=n-1)
    else:
        return x
    
tp.save(... # Not sure if you need to have a partial here.

**Review by achoum — APPROVED** [review]
This is awesome

File: temporian/core/test/serialization_test.py (hunk 6 above)
Comment (ianspektor) [reply temporian/core/test/serialization_test.py:521]: great suggestions! done

**Pull request merged**